### PR TITLE
llama : add Deepseek support #5981

### PR DESCRIPTION
**Pull request by dragnil1**
ref #5981 

- Work in progress. Creating this pr to check If I am on track with the possible implementation https://github.com/ggerganov/llama.cpp/pull/5464#issuecomment-1974818993

### CODE REVIEW

File: unicode.h [hunk 1]
```diff
@@ -24,3 +26,20 @@ int unicode_cpt_type(const std::string & utf8);
 std::string unicode_byte_to_utf8(uint8_t byte);
 uint8_t unicode_utf8_to_byte(const std::string & utf8);
 
+std::vector<std::wstring> get_gpt2_regex();
+std::vector<std::wstring> get_deepseek_coder_regex();
+std::vector<std::wstring> get_deepseek_llm_regex();
```
Comment (ggerganov) [inline unicode.h:31]: I'm thinking the interface here should be:

```c++
std::vector<std::string> unicode_regex_split(const std::string & text, const std::vector<std::string> & regexes);
```

The implementation should be something like what `regex_bpe_preprocess` currently is. It loops through the regex strings and if we have a unicode representation (e.g. `"//\s?\p{L}+" -> std::wstring`) - we apply it with `std::wregex`. Else, if we have a custom implementation (for example see the GPT2 preprocess function) then we apply that.

The `unicode` model should not have any kind of notion about GPT2, Deepseek or other model-related stuff. This information should be in `llama.cpp`

File: llama.cpp [hunk 2]
```diff
@@ -10003,143 +10024,82 @@ struct llm_tokenizer_bpe {
...
+        return byte_encoding_process(bpe_words);
+    }
+
+    std::vector<std::string> bpe_gpt2_preprocess(const std::string & text) {
+        return regex_bpe_preprocess(text, get_gpt2_regex());
+    }
+
+    std::vector<std::string> bpe_deepseek_coder_preprocess(const std::string & text) {
+        return regex_bpe_preprocess(text, get_deepseek_coder_regex());
+    }
```
Comment (ggerganov) [inline llama.cpp:10099]: Following my previous comment, this should eventually become:

```suggestion
    std::vector<std::string> bpe_deepseek_coder_preprocess(const std::string & text) {
        return unicode_regex_split(text, {
	        "[\\p{P}\\$\\+<=>\\^~\\|]+",
	        "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)",
	        "[0-9][0-9][0-9]"
	        "\\s?\\p{L}+",
	        "\\s?\\p{P}+",
	        "\\p{N}",
        });
    }
```

File: unicode.h (hunk 1 above)
Comment (dragnil1) [reply unicode.h:31]: Regarding the custom implementaton, does the GPT2 preprocess function work for deepseek models? Or, we have implement the custom function for deepseek models for ourselves.

File: llama.cpp [hunk 3]
```diff
@@ -10003,151 +10024,41 @@ struct llm_tokenizer_bpe {
...
-    }
-
     const llama_vocab & vocab;
 
     std::vector<llm_symbol> symbols;
     std::vector<llm_symbol> symbols_final;
 
     llm_bigram_bpe::queue work_queue;
+
+    const std::vector<std::wstring> gpt2_regex = {
```
Comment (ggerganov) [inline llama.cpp:10034]: I had a different idea about this - let me try to explain again:

In `llama.cpp`, we want to keep the original regex strings as they have been specified by the model creators. For example:

- `'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)`
- `\s?\p{L}+`
- etc.

Now, my understanding is that in C++ we cannot simple perform some of those regex matching due to the lack of support for some the regex patterns in the standard library. So to solve this issue, we create the `unicode` module, which takes the regex strings from above as they are and performs a few different strategies to split the target string:

- If we have a known unicode representation generated in some way, we apply that using `std::wregex`. I.e. we check a constant `std::map<std::string, std::wstring>` for the presence of the regex
- If not, we then check if we have a custom implementation of the regex via a function call (see `bpe_gpt2_preprocess()` on master which is a custom implementation of regex split with `'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+`)
- Else, we just apply `std::regex` and hope for the best, or throw an error

Comment (dragnil1) [reply llama.cpp:10034]: Hello, in the `unicode.cpp`, I have implemented the `unicode_regex_split()` function which iterates through the given regexes and if some match is found, it uses the regex. Otherwise, it  uses the modified `bpe_gpt2_preprocess()` function renamed as `unicode_custom_preprocess()`. Now, I have some questions regarding `bpe_gpt2_preprocess()`. Can it handle the input of deepseek coder and deepseek llm? If no, so, I have to write custom function for them when regex is not found?

Comment (ggerganov) [reply llama.cpp:10034]: > Can it handle the input of deepseek coder and deepseek llm?

No, AFAIK it is not compatible with the deepseek regex. The way I understand it is that `bpe_gpt2_preprocess()` (i.e. `unicode_custom_preprocess()`) works only for the following regex (based on the comment in the code):

```
's|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+
```

So the logic in `unicode_regex_split()` has to check if this is the input regex and only apply this specific custom implementation in that case. For other regexes, we might want to implement more custom implementations in other functions and use them in `unicode_regex_split()` in the future.

Note that this my understanding of how this part of the tokenizer is supposed to work. I could be wrong, so don't take all of these suggestions for granted.

In any case, the huge unicode constants like `gpt2_regex` should not be located in `llama.cpp`, but instead should be in `unicode.cpp`.

**Comment by ggerganov** [comment]
@dragnil1 Thank you for the help. With LLaMA v3 now switching to BPE tokenizer, this functionality becomes very important (see https://github.com/ggerganov/llama.cpp/issues/6914).

I'll focus on finalizing it ASAP and will likely use the work in this PR as a starting point. Will probably move the branch in the `llama.cpp` repo so that we can run `ggml-ci` as well.

The tokenization and unicode handling is definitely not my strongest and favourite part of the codebase, so if you or anyone else have any insights, don't hesitate to share or help out. I think I have the understanding of how to implement BPE pre-processing support, but I could very well be missing something

**Comment by dragnil1** [comment]
> @dragnil1 Thank you for the help. With LLaMA v3 now switching to BPE tokenizer, this functionality becomes very important (see #6914).
> 
> I'll focus on finalizing it ASAP and will likely use the work in this PR as a starting point. Will probably move the branch in the `llama.cpp` repo so that we can run `ggml-ci` as well.
> 
> The tokenization and unicode handling is definitely not my strongest and favourite part of the codebase, so if you or anyone else have any insights, don't hesitate to share or help out. I think I have the understanding of how to implement BPE pre-processing support, but I could very well be missing something

Thanks for letting me work on this pr. Sorry for the delay. I was working on it to get it work on windows. The recent commit passed the tests in ubuntu. But the tokenizer tests failed on windows. I had the idea of using std::wregex when the wchar_t size is 32 bits and std::regex when wchart_t size is 16. But using regex is giving SEGFAULT on tokenizer tests on windows. This is probably because the regex pattern from ReFlex library is too large than the the regex pattern used for wregex. While doing some research on it, I found the most efficient way will be to use boost library with icu support. But it will hamper the minimal dependency support of llama.cpp. Otherwise, we can use the standalone boost regex library.

**Comment by ggerganov** [comment]
> But using regex is giving SEGFAULT on tokenizer tests on windows. This is probably because the regex pattern from ReFlex library is too large than the the regex pattern used for wregex.

In the latest version #6920, I changed the order of the the regexes: first look for custom implementation and then look for a known equivalent regex. I also disabled DeepSeek code-paths temporary until we get the tests running as they were on `master`. So with these changes, I don't think we apply large regexes, but it still crashes (based on the Windows CIs).

I don't have a Windows environment to work on, so it's gonna take me some time to figure out where it goes wrong

**Comment by dragnil1** [comment]
Ok. I have found the reason of the test failing in windows. Some regex ranges are not valid in windows.  Here is an example which **will run**.
```
#include <iostream>
#include <string>
#include <regex>

int main() {
    std::wregex pattern2(L"[\U00000041-\U0000005A]"); // will run
    return 0;
}
```
Here is an example which **will not run**.
```
#include <iostream>
#include <string>
#include <regex>

int main() {
    std::wregex pattern1(L"[\U00011700-\U0001171A]"); // will not run
    return 0;
}
```

**Comment by ggerganov** [comment]
Yes, I just noticed the error in the CI:

```
6: llama_model_load: error loading model: error loading model vocabulary: regex_error(error_range): The expression contained an invalid character range, such as [b-a] in most encodings.
```

https://github.com/ggerganov/llama.cpp/actions/runs/8850389799/job/24304574061?pr=6920#step:12:1392

Any ideas how to resolve?

**Comment by dragnil1** [comment]
> Yes, I just noticed the error in the CI:
> 
> ```
> 6: llama_model_load: error loading model: error loading model vocabulary: regex_error(error_range): The expression contained an invalid character range, such as [b-a] in most encodings.
> ```
> 
> https://github.com/ggerganov/llama.cpp/actions/runs/8850389799/job/24304574061?pr=6920#step:12:1392
> 
> Any ideas how to resolve?

We cannot use ranges which contain codepoints that requires greater than 2 bytes. We have to convert the 3 bytes or 4 bytes ranges to individual values. But this may cause the regex pattern to be large enough to result in a SEGFAULT.
